File: ios/Stockbook/Features/Book/ExpenseSheet.swift
import SwiftUI

/// Writing down money the owner spent, and correcting it later.
///
/// Two fields and a date. There is no number to type, unlike every other
/// document in this app: an invoice, a receipt and a credit note all have a
/// number because there is a slip in a drawer carrying the same one, and there
/// is no such slip behind a tank of petrol.
///
/// Removing is a plain ghost button with no confirmation, where removing a bill
/// asks twice. That is not carelessness about the owner's data — it is that
/// nothing else moves. A deleted bill puts stock back and frees a number; this
/// is a line leaving a private list, and it can be typed again in ten seconds.
struct ExpenseSheet: View {
    @Environment(StockbookStore.self) private var store
    @Environment(\.currency) private var currency

    /// The expense being corrected, or nil for a new one.
    var editing: Expense?
    let onClose: () -> Void

    @State private var amount = ""
    @State private var note = ""
    @State private var spentAt = Date.now

    private var typed: Double { Money.parse(amount) ?? 0 }
    private var canSave: Bool { typed > 0 && !note.isBlank }

    /// What the shop has called an expense before, filtered by what is typed.
    ///
    /// Dropped entirely once the box matches one exactly — offering somebody the
    /// word they have just finished typing is a row that can only be in the way.
    private var suggestions: [String] {
        let found = store.expenseNotes(matching: note)
        if found.count == 1, found[0].lowercased() == note.trimmed.lowercased() { return [] }
        return found
    }

    private func choose(_ suggestion: String) {
        note = suggestion
        dismissKeyboard()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            SheetHeader(
                title: editing == nil ? Loc.newExpense : Loc.editExpense,
                onClose: onClose
            )

            NocturneField(
                label: Loc.expenseWhatFor,
                placeholder: Loc.expenseWhatForHint,
                text: $note,
                isRequiredAndEmpty: note.isBlank,
                identifier: "expense.note"
            )

            // What the shop keeps buying, most-used first. A shortcut and never
            // a requirement: unlike the customer picker, where a typed name has
            // no account behind it, anything typed here is a perfectly good
            // expense — so the list only ever saves keystrokes.
            //
            // Hidden the moment the box already says exactly what a suggestion
            // says, or it sits there offering the owner the word they have just
            // finished typing.
            if !suggestions.isEmpty {
                VStack(spacing: 0) {
                    ForEach(suggestions, id: \.self) { suggestion in
                        Button { choose(suggestion) } label: {
                            HStack(spacing: 8) {
                                Glyph(Icon.expenses, size: 12)
                                    .foregroundStyle(Nocturne.neutral500)
                                Text(suggestion)
                                    .font(NocturneType.inter(13.5))
                                    .frame(maxWidth: .infinity, alignment: .leading)
                                    .lineLimit(1)
                            }
                            .padding(.horizontal, 11)
                            .frame(height: 34)
                            .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                    }
                }
                .padding(.vertical, 3)
                .frame(maxWidth: .infinity)
                .background(Nocturne.surface)
                .clipShape(RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous))
                .hairline(Nocturne.accent, radius: Metrics.controlRadius)
                .padding(.top, 6)
            }

            Color.clear.frame(height: 10)

            HStack(alignment: .bottom, spacing: 8) {
                NocturneField(
                    label: Loc.amountField,
                    placeholder: "0",
                    text: $amount,
                    height: 40,
                    keyboard: .decimalPad,
                    isRequiredAndEmpty: typed <= 0,
                    prefix: currency.symbol.trimmed,
                    fontSize: 13.5,
                    identifier: "expense.amount"
                )
                NocturneDateField(
                    label: Loc.expenseSpentOn,
                    date: $spentAt,
                    identifier: "expense.spentAt"
                )
            }

            Text(Loc.expensesArePrivate)
                .nocturneText(.meta)
                .padding(.top, 8)

            // The button says what is missing rather than going dead and silent —
            // the same rule every other sheet here follows.
            Button(saveTitle) { save() }
                .buttonStyle(PrimaryButtonStyle(fullWidth: true))
                .disabled(!canSave)
                .padding(.top, 14)

            if let editing {
                Button(Loc.removeExpense) {
                    store.deleteExpense(id: editing.id)
                    onClose()
                }
                .buttonStyle(GhostButtonStyle(tint: Nocturne.neutral500))
                .frame(maxWidth: .infinity)
                .padding(.top, 6)

                Text(Loc.removeExpenseNote)
                    .nocturneText(.meta)
                    .padding(.top, 4)
            }
        }
        .keyboardDoneButton()
        .onAppear(perform: fill)
    }

    private var saveTitle: String {
        if note.isBlank { return Loc.enterWhatItWasFor }
        if typed <= 0 { return Loc.enterAnAmount }
        return editing == nil ? Loc.saveExpense : Loc.saveChanges
    }

    /// Filled here rather than in the property initialisers, which cannot see
    /// `editing` — the same shape every other editing sheet in this app uses.
    private func fill() {
        guard let editing else { return }
        amount = Money.amount(editing.amount, in: currency)
        note = editing.note
        spentAt = editing.spentAt
    }

    private func save() {
        guard canSave else { return }
        if let editing {
            store.updateExpense(id: editing.id, amount: typed, note: note, spentAt: spentAt)
        } else {
            store.addExpense(amount: typed, note: note, spentAt: spentAt)
        }
        onClose()
    }
}

File: android/app/src/main/kotlin/com/stockbook/app/feature/book/ExpenseSheet.kt
package com.stockbook.app.feature.book

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.DatePicker
import androidx.compose.material3.DatePickerDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Text
import androidx.compose.material3.rememberDatePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.stockbook.app.design.DateField
import com.stockbook.app.design.GhostButton
import com.stockbook.app.design.Glyph
import com.stockbook.app.design.Icon
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.NocturneField
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.PrimaryButton
import com.stockbook.app.design.SheetHeader
import com.stockbook.app.design.card
import com.stockbook.app.design.hairline
import com.stockbook.core.model.Expense
import com.stockbook.core.model.ShopState
import com.stockbook.core.model.Timestamps
import com.stockbook.core.money.Money
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.Strings
import java.time.Instant
import java.time.ZoneId
import java.time.ZoneOffset

/**
 * Writing down money the owner spent, and correcting it later.
 *
 * Two fields and a date. There is no number to type, unlike every other document
 * in this app: an invoice, a receipt and a credit note all have a number because
 * there is a slip in a drawer carrying the same one, and there is no such slip
 * behind a tank of petrol.
 *
 * Removing is a plain ghost button with no confirmation, where removing a bill
 * asks twice. That is not carelessness about the owner's data — it is that
 * nothing else moves. A deleted bill puts stock back and frees a number; this is
 * a line leaving a private list, and it can be typed again in ten seconds.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ExpenseSheet(
    /** The expense being corrected, or null for a new one. */
    editing: Expense?,
    state: ShopState,
    store: StockbookStore,
    strings: Strings,
    onClose: () -> Unit
) {
    val currency = state.settings.currency

    var amount by remember(editing) {
        mutableStateOf(editing?.let { Money.amount(it.amount, currency) } ?: "")
    }
    var note by remember(editing) { mutableStateOf(editing?.note ?: "") }
    var spentAt by remember(editing) { mutableStateOf(editing?.spentAt ?: Timestamps.now()) }
    var pickingDate by remember { mutableStateOf(false) }

    val typed = Money.parse(amount) ?: 0.0
    val canSave = typed > 0 && note.isNotBlank()

    // What the shop has called an expense before, filtered by what is typed.
    // Keyed on `state` as well as the text: a plain call over a StateFlow's
    // current value subscribes to nothing, so the list would sit still after the
    // expense that should have joined it was saved.
    //
    // Dropped entirely once the box matches one exactly — offering somebody the
    // word they have just finished typing is a row that can only be in the way.
    val suggestions = remember(state, note) {
        store.expenseNotes(note).let { found ->
            if (found.size == 1 && found.single().equals(note.trim(), ignoreCase = true)) emptyList()
            else found
        }
    }

    Column(modifier = Modifier.fillMaxWidth()) {
        SheetHeader(
            title = if (editing == null) strings.newExpense else strings.editExpense,
            onClose = onClose
        )

        NocturneField(
            value = note,
            onValueChange = { note = it },
            label = strings.expenseWhatFor,
            placeholder = strings.expenseWhatForHint,
            isRequiredAndEmpty = note.isBlank(),
            modifier = Modifier.fillMaxWidth()
        )

        // A shortcut and never a requirement: unlike the customer picker, where a
        // typed name has no account behind it, anything typed here is a perfectly
        // good expense — so this list only ever saves keystrokes.
        if (suggestions.isNotEmpty()) {
            Spacer(Modifier.height(6.dp))
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .card(Metrics.controlRadius)
                    .hairline(Nocturne.accent, Metrics.controlRadius)
                    .padding(vertical = 3.dp)
            ) {
                suggestions.forEach { suggestion ->
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable { note = suggestion }
                            .padding(horizontal = 11.dp, vertical = 8.dp)
                    ) {
                        Glyph(Icon.expenses, size = 12.dp, tint = Nocturne.neutral500)
                        Spacer(Modifier.width(8.dp))
                        Text(
                            suggestion,
                            style = NocturneType.inter(13.5),
                            color = Nocturne.text,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                    }
                }
            }
        }
        Spacer(Modifier.height(10.dp))

        Row(modifier = Modifier.fillMaxWidth()) {
            NocturneField(
                value = amount,
                onValueChange = { amount = it },
                label = strings.amountField,
                placeholder = "0",
                numeric = true,
                prefix = currency.symbol.trim(),
                isRequiredAndEmpty = typed <= 0,
                height = 40.dp,
                fontSize = 13.5,
                modifier = Modifier.weight(1f)
            )
            Spacer(Modifier.width(8.dp))
            DateField(
                label = strings.expenseSpentOn,
                value = strings.pickedDate(spentAt),
                onClick = { pickingDate = true },
                height = 40.dp,
                modifier = Modifier.weight(1f)
            )
        }

        if (pickingDate) {
            val picker = rememberDatePickerState(initialSelectedDateMillis = spentAt.toEpochMilli())
            DatePickerDialog(
                onDismissRequest = { pickingDate = false },
                confirmButton = {
                    GhostButton(strings.done, onClick = {
                        picker.selectedDateMillis?.let { millis ->
                            // The picker hands back midnight UTC. Re-anchored to
                            // midday in the phone's own zone, so the expense lands
                            // on the day the owner tapped whatever the offset —
                            // which is what `spentIn` buckets by.
                            spentAt = Instant.ofEpochMilli(millis)
                                .atZone(ZoneOffset.UTC)
                                .toLocalDate()
                                .atTime(12, 0)
                                .atZone(ZoneId.systemDefault())
                                .toInstant()
                        }
                        pickingDate = false
                    })
                }
            ) {
                DatePicker(state = picker)
            }
        }

        Spacer(Modifier.height(8.dp))
        Text(
            strings.expensesArePrivate,
            style = NocturneType.meta,
            color = Nocturne.neutral500
        )

        Spacer(Modifier.height(14.dp))
        PrimaryButton(
            // The button says what is missing rather than going dead and silent —
            // the same rule every other sheet here follows.
            title = when {
                note.isBlank() -> strings.enterWhatItWasFor
                typed <= 0 -> strings.enterAnAmount
                editing == null -> strings.saveExpense
                else -> strings.saveChanges
            },
            onClick = {
                if (editing == null) {
                    store.addExpense(typed, note, spentAt)
                } else {
                    store.updateExpense(editing.id, typed, note, spentAt)
                }
                onClose()
            },
            enabled = canSave,
            fullWidth = true,
            modifier = Modifier.fillMaxWidth()
        )

        if (editing != null) {
            Spacer(Modifier.height(6.dp))
            GhostButton(
                strings.removeExpense,
                onClick = {
                    store.deleteExpense(editing.id)
                    onClose()
                },
                tint = Nocturne.neutral500,
                modifier = Modifier.fillMaxWidth()
            )
            Spacer(Modifier.height(4.dp))
            Text(
                strings.removeExpenseNote,
                style = NocturneType.meta,
                color = Nocturne.neutral500,
                modifier = Modifier.fillMaxWidth().padding(bottom = 2.dp)
            )
        }
    }
}
